chill-scala/src/test/scala/com/twitter/chill/ClosureCleanerSpec.scala

```diff
@@ -16,59 +16,227 @@ limitations under the License.
+      serializableFn(closure2, before = false, after = true)
+    }
+
+    "clean basic closures (LMF)" in {
+      assume(ClosureCleanerSpec.supportsSerializedLambda)
+
+      val closure1 = (_: Int) => someSerializableValue
+      val closure2 = (_: Int) => someSerializableMethod()
+
+      serializableFn(closure1, before = false, after = false)
+      serializableFn(closure2, before = false, after = false)
+    }
+
+    "clean functions in traits" in {
+      // not serializable; AwesomeFn2 needs to extend Serializable
+      val fn = BaseFns2.timesByMult
+      serializableFn(fn, before = false, after = false)
+    }
+
+    "clean basic nested closures (non LMF)" in {
+      assume(!ClosureCleanerSpec.supportsSerializedLambda)
+
+      val closure1 = (i: Int) => {
+        Option(i).map { x =>
+          x + someSerializableValue
+        }
+      }
+      val closure2 = (j: Int) => {
+        Option(j).map { x =>
+          x + someSerializableMethod()
+        }
+      }
+      val closure3 = (m: Int) => {
+        Option(m).foreach { x =>
+          Option(x).foreach { y =>
+            Option(y).foreach { z =>
+              someSerializableValue
+            }
+          }
+        }
       }
 
-      ClosureCleaner(myFun)
-      myFun(1) should equal(2)
-      myFun(2) should equal(4)
+      serializableFn(closure1, before = false, after = true)
+      serializableFn(closure2, before = false, after = true)
+      serializableFn(closure3, before = false, after = true)
+    }
 
-      case class Test(x: Int)
-      val t = Test(3)
-      ClosureCleaner(t)
-      t should equal(Test(3))
+    "clean basic nested closures (LMF)" in {
+      assume(ClosureCleanerSpec.supportsSerializedLambda)
+
+      val closure1 = (i: Int) => {
+        Option(i).map { x =>
+          x + someSerializableValue
+        }
+      }
+      val closure2 = (j: Int) => {
+        Option(j).map { x =>
+          x + someSerializableMethod()
+        }
+      }
+      val closure3 = (m: Int) => {
+        Option(m).foreach { x =>
+          Option(x).foreach { y =>
+            Option(y).foreach { z =>
+              someSerializableValue
+            }
+          }
+        }
+      }
+
+      serializableFn(closure1, before = false, after = false)
+      serializableFn(closure2, before = false, after = false)
+      serializableFn(closure3, before = false, after = false)
     }
 
-    "handle outers with constructors" in {
+    "clean outers with constructors" in {
       class Test(x: String) {
-        val l = x.size
-        def rev(y: String) = (x + y).size
+        val l: Int = x.length
+        def rev(y: String): Int = (x + y).length
       }
 
       val t = new Test("you all everybody")
-      val fn = t.rev _
-      //debug(fn)
-      //println(ClosureCleaner.getOutersOf(fn))
-      ClosureCleaner(fn)
-      fn("hey") should equal(20)
+      val fn: String => Int = t.rev
+      serializableFn(fn, before = false, after = false)
```

Comment:
there is no use of ClosureCleaner in this test. Does it not work with this case?

Reply:
ClosureCleaner is applied in serializableFn (maybe we should change the name?) yeah this is a strange case i can't manage to find an $outer reference in either 2.11 and 2.12